File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations
from math import floor

def solution(numbers):
    answer = 0
    d = dict()
    
    for n in range(1, len(numbers) + 1):
        for k in permutations(numbers, n):
            if check(d, k):
                answer += 1
    
    return answer

def check(d, numbers):
    num = int("".join(numbers))
    if num in d: return False
    d[num] = 1
    return is_prime(num)


def is_prime(n):
    if n < 2: return False
    if n <= 3: return True
    if n % 2 == 0: return False
    for i in range(3, floor(n ** 0.5) + 1, 2):
        if n % i == 0: return False
    return True

File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    public int solution(String numbers) {
        Set<Integer> primes = new HashSet();
        check("", Arrays.asList(numbers.split("")), primes);
        return primes.size();
    }
    
    private void check(String n, List<String> numbers, Set<Integer> primes){
        if(n.length() > 0){
            int num = Integer.valueOf(n);
            if(isPrime(num)) primes.add(num);    
        }
        if(numbers.isEmpty()) return;
        for(int i = 0; i < numbers.size(); i++){
            String newChar = n + numbers.get(i);
            List<String> newNumbers = new ArrayList(numbers);
            newNumbers.remove(i);
            check(newChar, newNumbers, primes);
        }
    }
    
    private boolean isPrime(int n){
        if(n <= 1) return false;
        if(n <= 3) return true;
        for(int i = 2; i <= Math.floor(Math.sqrt(n)); i++){
            if(n % i == 0) return false;
        }
        return true;
    }
}

/*
상태
(str, numbers, primes)

종료 조건
numbers.isEmpty()

점화식
(str, numbers, primes) ->
(str + numbers[0], numbers - numbers[0], primes)
(str + numbers[1], numbers - numbers[1], primes)
...


*/
